core/trino-main/src/main/java/io/trino/sql/planner/SplitSourceFactory.java

```diff
@@ -175,9 +179,29 @@ private Map<PlanNodeId, SplitSource> visitScanAndFilter(TableScanNode node, Opti
             }
 
             Constraint constraint = filterPredicate
-                    .map(predicate -> filterConjuncts(plannerContext.getMetadata(), predicate, expression -> !DynamicFilters.isDynamicFilter(expression)))
-                    .map(predicate -> new LayoutConstraintEvaluator(plannerContext, typeAnalyzer, session, typeProvider, node.getAssignments(), predicate))
-                    .map(evaluator -> new Constraint(TupleDomain.all(), evaluator::isCandidate, evaluator.getArguments())) // we are interested only in functional predicate here, so we set the summary to ALL.
+                    .map(predicate -> {
+                        Metadata metadata = plannerContext.getMetadata();
+                        predicate = filterConjuncts(metadata, predicate, expression -> !isDynamicFilter(expression) && isDeterministic(expression, metadata));
+                        DomainTranslator.ExtractionResult decomposedPredicate = DomainTranslator.getExtractionResult(plannerContext, session, predicate, typeProvider);
+                        ConnectorExpressionTranslator.ConnectorExpressionTranslation expressionTranslation = ConnectorExpressionTranslator.translateConjuncts(
+                                session,
+                                decomposedPredicate.getRemainingExpression(),
+                                typeProvider,
+                                plannerContext,
+                                typeAnalyzer);
+                        Map<String, ColumnHandle> connectorExpressionAssignments = node.getAssignments()
+                                .entrySet().stream()
+                                .collect(toImmutableMap(entry -> entry.getKey().getName(), Map.Entry::getValue));
+                        // TODO Constraint.predicate is supposed not to cover Constraint.expression, but not Constraint.summary
+                        LayoutConstraintEvaluator evaluator = new LayoutConstraintEvaluator(plannerContext, typeAnalyzer, session, typeProvider, node.getAssignments(), predicate);
```

Comment:
`PushPredicateIntoTableScan` mentions that this could become expensive and guards using `LayoutConstraintEvaluator` with a boolean flag `pruneWithPredicateExpression` which is only set to true in one usage I can find. It's also checks `!TRUE_LITERAL.equals(decomposedPredicate.getRemainingExpression())`. Should we do something similar here?

Reply:
> `PushPredicateIntoTableScan` mentions that this could become expensive and guards using `LayoutConstraintEvaluator` with a boolean flag `pruneWithPredicateExpression` which is only set to true in one usage I can find

correct
`PushPredicateIntoTableScan` is run many times for a query, but `pruneWithPredicateExpression` is currently set only once (in AddExchanges afair), so `O(# table scan)` times.

i think doing it once more time here is OK. anyway, this is pre-existing, the code on master already has `LayoutConstraintEvaluator` and what we're doing here doesn't add more work for it

> . It's also checks !TRUE_LITERAL.equals(decomposedPredicate.getRemainingExpression()). Should we do something similar here?

probably yes. Note however the TODO there. I didn't want to regress on Iceberg in this PR, so doing this better would require some connector side change as well.